tests/test_const_ref_caster.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+    tests/test_const_ref_caster.cpp
```

Comment:
These tests seem more appropriate in `test_builtin_casters.cpp`/`test_builtin_casters.py`.

Reply:
Perhaps, but I think that since they are intended to exercise different aspects, putting it into a separate file makes the intent a bit more clear.

Reply:
Yes, pybind11's tests are a mess. But I think the idea it to keep a minimal amount of files, because they all take quite some time to compile. I also don't think this one test is enough to create a separate file for.

Meanwhile, `type_caster<std::reference_wrapper<T>>` is already tested in `test_builtin_casters.cpp` and other test files already have custom casters as well (`test_custom_type_casters.cpp`, `test_copy_move.cpp`, and even one in `pybind11_tests.h`). So if we can keep the scope of the caster focused enough, I think it's fine.

Reply:
The other counter point is that this requires a custom type caster; so maybe move them to custom_type_casters.cc?  I think that the current test is about as minimal as it can be while exercising the various paths we want to test.

Reply:
Just logging my view, not a request: it makes me sad to see this test lumped in with other things. I know the existing tests do that a lot, but I think it really isn't a good approach, because it convolutes unrelated things, makes them harder to understand/follow, and obviously is counter the the idea of testing small self-contained units. Sticking with the established pattern only has so much value. Nudging things in a better direction has value, too. My preference is definitely to keep this test separate, if Boris or Yannick can agree.

Reply:
Anyway, Done.

Reply:
I agree with @rwgk however I will change it to the reviewers consensus.

Reply:
> The other counter point is that this requires a custom type caster; so maybe move them to custom_type_casters.cc? I think that the current test is about as minimal as it can be while exercising the various paths we want to test.

Well, coming from this PR, it's testing the built-in caster for `std::reference_wrapper`, so I thought it would be best to keep this as close to the other tests for `reference_wrapper`, and to not create yet another file is "something with casters" (which takes time to compile, each time!).

> move them to custom_type_casters.cc

It's not really about the custom caster, though. It's about testing `reference_wrapper`, currently, I though? If this maybe gets expanded further in the future, the caster could be shared in a header, though, when it's necessary.

> it makes me sad to see this test lumped in with other things.

Why exactly? Because I had explicitly tried to design/shrink this to be as minimal, unit-testy as possible, such that it could fit in with the rest of the `std::reference_wrapper` tests (that were even extended) in `builtin_casters.cpp`.

And to me it's more confusing to break convention than to add a few lines to the existing tests.

I do agree the tests are messy, btw, but I don't see the urgency of splitting things up. Build times are (partially) to blame, I think. So maybe once #2445 gets merged, it will be more feasible to split up tests, in a separate PR that brings order to the tests in a less ad-hoc fashion?

> I agree with @rwgk however I will change it to the reviewers consensus.

Not sure there's consensus, though. If you want, we can give @henryiii a decisive vote, but ... well, yeah, I've explained my reasoning :-)